server/src/utils/param_parser.rs

```diff
@@ -0,0 +1,195 @@
+use serde_json::Value;
+
+enum IndexingState {
+    LookingForLeftParenthesis,
+    Accumulating(usize),
+}
+
+// updates a json Value at the specified path
+pub fn update_json_object<'a, V: AsRef<str>>(obj: &'a mut Value, path: &[&str], val: V) {
+    update_json_object_internal(obj, path, val, false)
+}
+
+fn update_json_object_internal<'a, V: AsRef<str>>(
+    obj: &'a mut Value,
+    path: &[&str],
+    val: V,
+    recursion_protection: bool,
+) {
+    if path.len() == 0 {
+        return;
+    }
+    if let Value::Object(the_map) = obj {
+        if path.len() == 1 {
+            the_map.insert(path[0].into(), Value::String(val.as_ref().to_string()));
+        } else {
+            match the_map.get_mut(path[0]) {
+                Some(inner) => {
+                    if inner.is_object() {
+                        // we have a member at the specified index and it is an object (we can insert at path)
+                        update_json_object_internal(inner, &path[1..], val, false);
+                    }
+                }
+                None => {
+                    if recursion_protection {
```

Comment:
Is it possible to write this without recursion protection? It seems like this is basically doing:

```rust
let map = match obj { Value::Object(map) => map, _ => return };
let (key, rest) = match path.split_first() { Some(tuple) => tuple, None => return };

let entry = map.entry(key.into());
if rest.is_empty() {
  entry.or_insert_with(|| val.to_string().into());
} else {
  update_json_object(entry.or_insert_with(|| Value::Object(Map::new())));
}

```

Reply:
Recursion protection is added just against programming bugs.
The code works without this, I like to write it this way when I do a recursive call without consuming part of the arguments (i.e. I modify some state and I do a recursive call with the exact same parameters), so that if there is a bug it is very clear why I have an endless recursion.

If you don't like it we can simply remove it.

Reply:
Without caching infinite loops (which should not happen) the function becomes:

```rust
/// Updates a json Value at the specified path.
pub fn update_json_object<'a, V: AsRef<str>>(obj: &'a mut Value, path: &[&str], val: V) {
    if path.len() == 0 {
        return;
    }
    if let Value::Object(the_map) = obj {
        if path.len() == 1 {
            the_map.insert(path[0].into(), Value::String(val.as_ref().to_string()));
        } else {
            match the_map.get_mut(path[0]) {
                Some(inner) => {
                    if inner.is_object() {
                        // we have a member at the specified index and it is an object (we can insert at path)
                        update_json_object(inner, &path[1..], val);
                    }
                }
                None => {
                    //nothing yet at the specified path create an object
                    the_map.insert(path[0].into(), Value::Object(serde_json::Map::new()));
                    // now we should have an object at the path, try again
                    update_json_object(obj, path, val);
                }
            }
        }
    }
}
```